experimental/web/sample_webgpu/main.c

```diff
@@ -546,16 +551,30 @@ static iree_status_t print_outputs_from_call(
         }
       }
     } else if (iree_vm_variant_is_ref(variant)) {
-      if (iree_hal_buffer_view_isa(variant.ref)) {
-        iree_hal_buffer_view_t* buffer_view =
-            iree_hal_buffer_view_deref(variant.ref);
-        // TODO(scotttodd): join async outputs together and return to caller
-        iree_hal_device_t* device = iree_runtime_session_device(call->session);
-        IREE_RETURN_IF_ERROR(print_buffer_view(device, buffer_view));
-      } else {
-        IREE_RETURN_IF_ERROR(iree_string_builder_append_cstring(
-            outputs_builder, "(no printer)"));
+      // Interpret the mapped buffer in the same format as the output view.
+      iree_hal_buffer_view_t* heap_buffer_view = NULL;
+      IREE_RETURN_IF_ERROR(iree_hal_buffer_view_create_like(
+          call_state->output_states[i].mapped_host_buffer,
+          call_state->output_states[i].buffer_view, iree_allocator_system(),
+          &heap_buffer_view));
+      // Query total length (excluding NUL terminator).
+      iree_host_size_t result_length = 0;
+      iree_status_t status = iree_hal_buffer_view_format(
+          heap_buffer_view, SIZE_MAX, 0, NULL, &result_length);
+      if (!iree_status_is_out_of_range(status)) return status;
+      status = iree_ok_status();
+      ++result_length;  // include NUL
+      // Format into a scratch buffer and append.
+      char* scratch = (char*)iree_alloca(result_length);
```

Comment:
this is dangerous - alloca only really works for a small number of KB of allocations (dozens of bytes to 1-4KB)

instead, you can use iree_string_builder_append_inline - you compute the required length like you do, call that with the length+1 (NUL) to get the char* pointer, then iree_hal_buffer_view_format right into the storage.

Reply:
Even better, I can just call `iree_hal_buffer_view_append_to_builder` that already does that :)